sktime/transformers/rocket.py

```diff
@@ -0,0 +1,178 @@
+from numba import njit, prange
+import numpy as np
+import pandas as pd
+
+from .base import BaseTransformer
+
+__author__ = "Angus Dempster"
+__all__ = ["ROCKET"]
+
+class Rocket(BaseTransformer):
+    """ROCKET
+
+    RandOm Convolutional KErnel Transform
+
+    @article{dempster_etal_2019,
+      author  = {Dempster, Angus and Petitjean, Fran\c{c}ois and Webb, Geoffrey I},
+      title   = {ROCKET: Exceptionally fast and accurate time series classification using random convolutional kernels},
+      year    = {2019},
+      journal = {arXiv:1910.13051}
+    }
+
+    Parameters
+    ----------
+    input_length : int, length of input time series
+    num_kernels  : int, number of random convolutional kernels
+    num_channels : int, number of channels for multivariate time series (default 1)
+    """
+
+    def __init__(self,  input_length, num_kernels, num_channels = 1):
+        self.input_length = input_length
+        self.num_kernels = num_kernels
+        self.num_channels = num_channels
+        self.kernels = _generate_kernels(input_length, num_kernels, num_channels)
+
+    def fit(self, X, y = None):
+        return self
+
+    def fit_transform(self, X, y = None):
+        return self.fit(X).transform(X)
+
+    def transform(self, X):
+        """Transforms input time series using random convolutional kernels.
+
+        Parameters
+        ----------
+        X : pandas DataFrame, input time series (sktime format)
+
+        Returns
+        -------
+        pandas DataFrame, transformed features
+        """
+        return pd.DataFrame(_apply_kernels(self._to_numpy(X), self.kernels))
+
+    @staticmethod
+    def _to_numpy(X, a = None, b = None):
+        return np.stack(X.iloc[a:b].applymap(lambda cell : cell.to_numpy()).apply(lambda row : np.stack(row), axis = 1).to_numpy())
+
+@njit("Tuple((float32[:],int32[:],float32[:],int32[:],int32[:],int32[:],int32[:]))(int64,int64,int64)")
+def _generate_kernels(input_length, num_kernels, num_channels):
+
+    candidate_lengths = np.array((7, 9, 11), dtype = np.int32)
+    lengths = np.random.choice(candidate_lengths, num_kernels)
+
+    num_channel_indices = np.zeros(num_kernels, dtype = np.int32)
+    for i in range(num_kernels):
+        limit = min(num_channels, lengths[i])
+        num_channel_indices[i] = 2 ** np.random.uniform(0, np.log2(limit + 1))
+
+    channel_indices = np.zeros(num_channel_indices.sum(), dtype = np.int32)
+
+    weights = np.zeros(np.int32(np.dot(lengths.astype(np.float32), num_channel_indices.astype(np.float32))), dtype = np.float32)
+    biases = np.zeros(num_kernels, dtype = np.float32)
+    dilations = np.zeros(num_kernels, dtype = np.int32)
+    paddings = np.zeros(num_kernels, dtype = np.int32)
+
+    a1 = 0 # for weights
+    a2 = 0 # for channel_indices
+
+    for i in range(num_kernels):
```

Comment:
Have you had a look at parallelisation with joblib? We're not sure if introducing numba as an additional dependency is a good idea at the moment.

Reply:
Obviously, in terms of the project, this decision is up to you.  However, if this implementation of ROCKET doesn't use numba (or, maybe, something else roughly equivalent), it really can't be used for benchmarking.

Unless you have a fairly fancy CPU, or access to a CPU cluster, parallelisation is likely only going to be a secondary benefit of using numba.  The main benefit of using numba is in compiling the functions that are doing all the work.  (In our experiements, we basically didn't use parallelisation: each experiment / dataset / run was performed on a single CPU core / thread.)

Given the nature of the computations required for ROCKET, compilation is really fundamental to making a python implementation work.  (In this implementation here, numba performs compilation is invisibly at import, and provides a massive, massive speedup over uncompiled python / numpy syntax.)

More generally, although my experience with numba is relatively limited, I strongly recommend it.

Reply:
Ok, thanks, yes I've used numba before, as I've said, not really sure at this point, for the distance metrics we chose to write them in Cython and then parallelise with joblib.